homeassistant/components/camera/__init__.py

```diff
@@ -0,0 +1,1133 @@
+        re.compile(r'/api/camera_proxy/(?P<entity_id>[a-zA-Z\._0-9]+)'),
+        _proxy_camera_image,
+        require_auth=True)
+
+    # pylint: disable=unused-argument
+    def _proxy_camera_mjpeg_stream(handler, path_match, data):
+        """ Proxies the camera image as an mjpeg stream via the HA server.
+        This function takes still images from the IP camera and turns them
+        into an MJPEG stream.  This means that HA can return a live video
+        stream even with only a still image URL available.
+        """
+        entity_id = path_match.group('entity_id')
+
+        camera = None
+        if entity_id in component.entities.keys():
+            camera = component.entities[entity_id]
+
+        if camera:
+            camera.last_connected_address = handler.address_string()
+            message = "{0} started streaming to {1}".format(
+                camera.name, handler.address_string())
+
+            hass.bus.fire(
+                "camera_stream_started",
+                {
+                    "component": DOMAIN,
+                    ATTR_ENTITY_ID: entity_id,
+                    ATTR_FRIENDLY_LOG_MESSAGE: message})
+
+            try:
+                camera.is_streaming = True
+                camera.update_ha_state()
+
+                http = urllib3.PoolManager()
+                handler.request.sendall(bytes('HTTP/1.1 200 OK\r\n', 'utf-8'))
+                handler.request.sendall(bytes(
+                    'Content-type: multipart/x-mixed-replace; \
+                        boundary=--jpgboundary\r\n\r\n', 'utf-8'))
+
+                handler.request.sendall(bytes('--jpgboundary\r\n', 'utf-8'))
+
+                while True:
+                    headers = urllib3.util.make_headers()
+
+                    if camera.username and camera.password:
+                        headers = urllib3.util.make_headers(
+                            basic_auth=camera.username + ':' + camera.password)
+
+                    req = http.request(
+                        'GET',
+                        camera.still_image_url,
+                        headers=headers)
+
+                    headers_str = ''
+                    headers_str = (
+                        headers_str + 'Content-length: ' +
+                        str(len(req.data)) + '\r\n')
+                    headers_str = headers_str + 'Content-type: image/jpeg\r\n'
+                    headers_str = headers_str + '\r\n'
+
+                    handler.request.sendall(
+                        bytes(headers_str, 'utf-8') +
+                        req.data +
+                        bytes('\r\n', 'utf-8'))
+
+                    handler.request.sendall(
+                        bytes('--jpgboundary\r\n', 'utf-8'))
+
+            # This needs to be a catchall exception as we need to stop
+            # streaming on any failure otherwise this will keep running
+            # forever
+            # pylint: disable=broad-except
+            except Exception:
```

Comment:
We should be able to figure out which exceptions can happen to prevent a catch all:
 
 - requests.RequestException if an error occurs fetching the image
 - IOError if an error occurs writing the output